WooCommerce/Classes/ViewModels/InAppPurchases/UpgradeViewState.swift

```diff
@@ -0,0 +1,35 @@
+import Foundation
+
+enum UpgradeViewState: Equatable {
+    case loading
+    case loaded(WooWPComPlan)
+    case purchasing(WooWPComPlan)
+    case waiting(WooWPComPlan)
+    case completed(WooWPComPlan)
+    case prePurchaseError(PrePurchaseError)
+    case purchaseUpgradeError(PurchaseUpgradeError)
+
+    var shouldShowPlanDetailsView: Bool {
+        switch self {
+        case .loading, .loaded, .purchasing, .prePurchaseError:
+            return true
+        default:
+            return false
+        }
+    }
+}
+
+// MARK: - Equatable conformance
+extension UpgradeViewState {
+    static func ==(lhs: UpgradeViewState, rhs: UpgradeViewState) -> Bool {
+        switch (lhs, rhs) {
+        case (.loading, .loading):
+            return true
+        case (.prePurchaseError(let lhsError), .prePurchaseError(let rhsError)):
+            return lhsError == rhsError
+        default:
+            // TODO: Needs conformance for the rest of cases
+            return false
+        }
+    }
```

Comment:
Yes, by making `WooWPComPlan` equatable we need to make conform `WooPlan`, `WooPlanFeatureGroup`, and `WooPlanFeature` as well. I was expecting to get most of it synthetized automatically since all these types are mainly made with String primitives, but I saw all sorts of errors as you mention. 

I initially thought these came from `let wpComPlan: WPComPlanProduct`, since it's a protocol and we would need to compare the concrete instances for equality, not the protocol itself, so a way to make it compile was to declare Equatable as:

```
struct WooWPComPlan: Equatable {
    static func == (lhs: WooWPComPlan, rhs: WooWPComPlan) -> Bool {
        return lhs.hardcodedPlanDataIsValid == rhs.hardcodedPlanDataIsValid &&
               lhs.wooPlan == rhs.wooPlan &&
               lhs.wpComPlan.id == rhs.wpComPlan.id &&
               lhs.wpComPlan.displayName == rhs.wpComPlan.displayName &&
               lhs.wpComPlan.description == rhs.wpComPlan.description &&
               lhs.wpComPlan.displayPrice == rhs.wpComPlan.displayPrice
    }

    let wpComPlan: WPComPlanProduct
    let wooPlan: WooPlan
    let hardcodedPlanDataIsValid: Bool
}
```
I'm thinking we could just also compare the pointers between wpComPlan instances by using `===`, if these are the same then all its properties are the same as well, but at this point I wasn't sure if that's the correct way and there are too many pieces involved just for the initial intended change of removing the ServiceLocator reference, so I went ahead with the minimum change. I'll investigate this further when we write the rest of the tests!